### ISSUE TITLE
[SR-16009] MainActor static-let initialization not enforced on main thread.

### ISSUE THREAD

**Issue by kavon** [apple/swift#58270]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-16009      |
   |Radar             | rdar://problem/90436577         |
   |Original Reporter | @kavon      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

swift-driver version: 1.26.9 Apple Swift version 5.5 (swiftlang-1300.0.31.1 clang-1300.0.29.1)

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, Concurrency        |
   |Assignee          | @kavon      |
   |Priority          | Medium      |

   md5: f6b9b747f9f450df4eddd1f7c0fe8e4f

  </details>

**cloned from**:  
* #57549 

**Issue Description:**

```swift
@MainActor 
struct Bar {
  init() { 
    print("bar init is main", Thread.isMainThread) 
  } 
  func barCall() { 
    print("bar call is main", Thread.isMainThread) 
  } 
} 

@MainActor 
struct Foo { 
  @MainActor 
  static let global = Foo() 

  init() { 
    print("foo init is main", Thread.isMainThread) 
    let b = Bar() 
    b.barCall() 
  } 

  func fooCall() { 
    print("foo call is main", Thread.isMainThread) 
  } 

} 

Task.detached { 
  await Foo.global.fooCall() 
} 
// prints: 
// foo init is main false 
// bar init is main false 
// bar call is main false 
// foo call is main true
```

While `fooCall()` is enforced on the main thread, the initialization of the static variable is not, which breaks the contract of `@MainActor`.

**Comment by billandjessparis** [comment]
This is still present in Swift 5.8

**Comment by moreindirection** [comment]
This is still present in Swift 5.8

**Comment by AnthonyLatsis** [comment]
Updated revision: Swift version 5.9-dev (LLVM 91080e65df791d6, Swift c1d5118c21da49a)

**Comment by bradleymackey** [comment]
This is still present in Swift 5.9

**Comment by hborla** [comment]
Fixed by https://github.com/apple/swift/pull/69598 and https://github.com/apple/swift/pull/69616

**Comment by moreindirection** [comment]
This is great news - thank you for fixing! 

When will this fix be available? Swift 5.10?

**Comment by bradleymackey** [comment]
@moreindirection If #69607 gets merged it will be.
